docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,537 @@
+  return o.length; // Error: Object does not have a "length" getter.
+}
+```
+
+Since the first line will definitely throw if `o` is not a String, the only way
+the second line could execute is if `o` is a string. But type promotion fails to
+understand that.
+
+To get around these limitations, the analyzer used by IntelliJ and others also
+has "type propagation". This is more sophisticated than type promotion. But,
+since the language spec controls what errors and warnings are shown, type
+propagation is limited to be using for things like auto-complete suggestions.
+
+Maintaining two similar but different flow typing systems adds complexity to our
+tools and confuses users. ("Why are you showing me an error that a variable
+isn't a Foo, when auto-complete on that same variable shows me all of Foo's
+methods?")
+
+To remedy that, this proposal augments type promotion with smarter rules when
+variables can be promoted. That should avoid requiring annoying redundant casts
+in more places, and let us unify type promotion and propagation into a single
+coherent system. This proposal doesn't cover all possible extensions to type
+propagation, but it handles the simplest, most common cases.
+
+## How Type Promotion Works
+
+The language specifies type promotion out of three pieces:
+
+1.  **Certain expressions cause a fact to be known about a variable.** For
+    example, `o` is String deduces the fact "`o` must be a String".
+
+2.  **The scope where a fact is known to be true.** For example in:
+
+    ```dart
+    if (o is String) {
+      return o.length;
+    } else {
+      throw "Not a string.";
+    }
+    ```
+
+    The fact "`o` is a String" is known to be true inside the then clause of the
+    if statement, but not the else clause, and not after the end of the if
+    statement.
+
+3.  **Which variables can have their types promoted.** Some code may cause a
+    fact to no longer be true. In that case, we shouldn't promote it. A simple
+    example is:
+
+    ```dart
+    if (o is String) {
+      o = 123; // Not a string anymore!
+      return o.length;
+    }
+    ```
+
+    After the assignment, it's definitely no longer a true fact that `o` is a
+    String.
+
+    There are other, more subtle ways where a fact may be true, but the analysis
+    is too complex to do. Usually this relates to closures that modify a
+    variable and may or may not be called, like:
+
+    ```dart
+    int stringLength(Object o, takeCallback(callback())) {
+      closure() {
+        o = 123; // Not a string.
+      }
+
+      if (o is String) {
+        takeCallback(closure); // ???
+        return o.length;
+      }
+    }
+    ```
+
+    After the invocation of `takeCallback()`, will `o` still be a String? Maybe.
+    It depends on whether the function passed to `stringLength()` calls the
+    closure that is passed to it. To avoid cases like this, the spec defines
+    certain variables to be off limits for type promotion.
+
```

Comment:
There is the 4th dimension: what type can be promoted (and how).

For example:
``` dart
Iterable<String> x = ...;
if (x is List) {
  ...
}
```

If we reduce the promotion to simple cases with only one constraint, then almost all cases are covered, and most users would be much happier. Dart2js seems to have some work in it.

@johnniwinther